### PR TITLE
Tuning training configs for SC integration

### PR DESCRIPTION
**Pull request by harimkang**
### This PR includes :
- Model & training schedule Config value update
- Related OTE PR: https://github.com/openvinotoolkit/training_extensions/pull/1150
- Prevent [data_cls] -> [data_cls] training when run without previous label schema in classification.

### CODE REVIEW

File: mpa/cls/stage.py [hunk 1]
```diff
@@ -206,20 +202,23 @@ def configure_task(cfg, training, model_meta=None, **kwargs):
...
-                    )
                     sampler_type = 'balanced'
+                    if len(model_classes) > 0:
+                        cfg.model.head.loss = ConfigDict(
+                            type='SoftmaxFocalLoss',
+                            loss_weight=1.0,
+                            gamma=gamma,
+                            reduction='none',
+                        )
+                    elif len(model_classes) == 0:
```
Comment (harimkang) [inline mpa/cls/stage.py]: @supersoob @JihwanEom  I think it's best to just use 'else'.

Comment (harimkang) [reply mpa/cls/stage.py]: Also, I don't think model_classes will always be zero for the Supervised Learning setting, what do you think?
For active learning and supervised learning, model_classes is expected to be greater than zero.

File: mpa/modules/hooks/early_stopping_hook.py [hunk 2]
```diff
@@ -62,12 +60,14 @@ def __init__(self,
                  rule: Optional[str] = None,
                  patience: int = 5,
                  iteration_patience: int = 500,
-                 min_delta: float = 0.0):
+                 min_delta: float = 0.0,
+                 warmup_iters: int = 0):
```
Comment (harimkang) [inline mpa/modules/hooks/early_stopping_hook.py]: @chuneuny-emily Is there a background in which this was added?

Comment (chuneuny-emily) [reply mpa/modules/hooks/early_stopping_hook.py]: It was required in mmseg default setting, but now it may not used with latest tuned parameters. After checking it, let me revert it to original one. Thanks for your feedback.

File: mpa/cls/stage.py (hunk 1 above)
Comment (supersoob) [reply mpa/cls/stage.py]: After reading your comment, I thought about the case that the supervised should be run. I'd say that when model classes and data classes are totally different without intersection, supervised learning needs to be initiated. Otherwise, should start incremental learning.
So I plan to change this part
`if len(set(model_classes) & set(dst_classes))==0: #change loss to cross entropy`
what do you think? have any better idea?
And I will change the default op into "REPLACE" to get the dst_classes correctly as you discussed in your previous PR.

Comment (harimkang) [reply mpa/cls/stage.py]: I've been thinking about this for a little bit..
For supervised learning, I think there are two cases:
- where model_classes and data_classes are the same
- or model_classes = []

```
if len(model_classes) == 0 or model_classes == data_classes:
  cfg.model.head.loss = dict(type='CrossEntropyLoss', loss_weight=1.0)
else:
  cfg.model.head.loss = ConfigDict(
      type='SoftmaxFocalLoss',
      loss_weight=1.0,
      gamma=gamma,
      reduction='none',
  )
```
In this case, I think you might have to look at the order of model_classes and data_classes (model_classes == data_classes).

There may be some parts that I missed, so please let me know if there are any.

Comment (supersoob) [reply mpa/cls/stage.py]: Agree that these two cases are for supervised learning.

- where model_classes and data_classes are the same -> No new classes thus not required incremental learning

- or model_classes = [] -> Model never been trained

And in my opinion, I think the case I mentioned in the last comment should be considered, too.

- len(set(model_classes) & set(dst_classes))==0 -> for example, the model has trained with some or large-scale data, but want to optimize it to my custom data

For example, there is a model trained with label [a,b], but the label schema from data is [c,d].
And they are totally different and no intersection here.
In that case, I'd say this is not incremental and the model is to run in the way of supervised learning/finetune.
Plus, since we already consented to enable REPLACE mode, the dst_classes will be replaced by [c,d] and the newly trained model's label schema will be turned to [c,d]. If this case is deemed as incremental, MERGE op should be set.

I don't know the correct answer about dealing the learning type, since there are so many cases and perspectives.
If you have any other opinion, please comment freely:)

Comment (harimkang) [reply mpa/cls/stage.py]: For our current use case, we can consider whether the label_schema changes or not.
- model_classes = [] -> SL
- model_classes == data_classes -> SL
- model_classes is not empty and label_schema changes -> C-IL
- but, [a, b] -> [c, d] case means you need to learn like SL, right?

Then, I think it's a good idea to check also whether it's MERGE or REPLACE. What do you think? (However, I think we need to find the case that is both REPLACE and Incremental.)


we need to write code to properly capture these all cases.

Comment (supersoob) [reply mpa/cls/stage.py]: - model_classes = [] -> SL from scratch **-> Yes**
- model_classes is not empty and label_schema changes -> C-IL **-> Yes, but may have a problematic case like you mentioned (REPLACE + Incremental)**
- but, [a, b] -> [c, d] case means you need to learn like SL, right? **-> Yes**

To clearly explain the third case,

-  model class = [a,b]
-  data class = [c,d]

Want to deal as incremental?
dst_classes = [a,b,c,d] (MERGE op required in current code base)

Want to deal as supervised?
dst_classes = [c,d] (REPLACE op required in current code base)

I think this should be deemed supervised learning case. Do you agree with it?

However, I found a problematic case in both REPLACE and Incremental mode you mentioned.
When model classes and data classes have same classes partially, like
model class = [a,b]
data class = [b,c]
dst_classes = [b,c] (REPLACE op)
dst_classes = [a,b,c] (MERGE op)

final dst_classes will be same as data classes when REPLACE op is enabled. If it learns in incremental mode, it would not learn some original model classes, which are not included in data classes (and actually it is not in data folder). In SC, since the data folder path is constant, we don't need to worry if data class is not a superset of model_class (or maybe not when classes are removed from the user). But when pre-trained model is loaded manually and partially has same classes with dataset, I wonder which learning mode the user wants (run incrementally? or from scratch with dataset?). If the case is for supervised learning, we can leave REPLACE op as default and need additional 'if condition' to deal with this loss change code. If not, MERGE op is needed only for this case. Which op need to choose? What do you think?

I hope @goodsong81 @JihwanEom @chuneuny-emily @jaegukhyun join this discussion and suggest ideas if you have time. If I missed something or understand in wrong way, please let me know or correct me.

Comment (harimkang) [reply mpa/cls/stage.py]: I think the most appropriate mode right now is 'REPLACE'. The selection of 'from scratch' can be made through the check box in the SC UI, so I don't think it's a consideration.

Rather, I think we can think about whether to take the 3rd case with incremental mode or supervision mode.

Comment (supersoob) [reply mpa/cls/stage.py]: Then disabling REPLACE and turning on MERGE is required for incremental learning in the problematic case I mentioned. I realized that it's orthogonal to the user's decision made by SC checkbox (whether scratch or not) in OTE code. Could you share your opinion about the third case or else?

File: mpa/modules/hooks/early_stopping_hook.py (hunk 2 above)
Comment (chuneuny-emily) [reply mpa/modules/hooks/early_stopping_hook.py]: Removed unnecessary code changes.

File: mpa/cls/stage.py (hunk 1 above)
Comment (supersoob) [reply mpa/cls/stage.py]: I changed the codes here reflecting this discussion. Thank you for giving meaningful comments:)

File: mpa/cls/stage.py [hunk 3]
```diff
@@ -297,17 +294,15 @@ def refine_tasks(train_cfg, meta, adapt_type):
     return model_tasks, old_tasks
 
 
-def refine_cls(train_cfg, meta, adapt_type):
+def refine_cls(train_cfg, data_classes, meta, adapt_type):
```
Comment (supersoob) [inline mpa/cls/stage.py:297]: I changed refine_cls and some codes to enable REPLACE mode in a right way. Before I changed this, REPLACE mode did not work at all in incremental learning. I ask for your review :)

File: models/segmentation/_base_/ocr_litehrnet18_mod2.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
+__norm_cfg = dict(type='BN', requires_grad=True)
+model = dict(
+    type='ClassIncrSegmentor',
+    is_task_adapt=True,
+    num_stages=1,
```
Comment (JihwanEom) [inline models/segmentation/_base_/ocr_litehrnet18_mod2.py]: Can we use EncoderDecoder class if we use only FCN head?

File: models/segmentation/_base_/ocr_litehrnet18_mod2.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+                      sampler=dict(type='MaxPoolingPixelSampler', ratio=0.25, p=1.7),
+                      loss_weight=1.0),
+             ]),
+    ],
+    train_cfg=dict(
+        mix_loss=dict(
+            enable=False,
+            weight=0.1
+        ),
+        loss_reweighting=dict(
```
Comment (JihwanEom) [inline models/segmentation/_base_/ocr_litehrnet18_mod2.py]: I'm not sure this loss_reweighting config is working but I believe this should be removed.

File: models/segmentation/_base_/ocr_litehrnet18_mod2.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+             enable_out_norm=False,
+             loss_decode=[
+                 dict(type='CrossEntropyLoss',
+                      use_sigmoid=False,
+                      sampler=dict(type='MaxPoolingPixelSampler', ratio=0.25, p=1.7),
+                      loss_weight=1.0),
+             ]),
+    ],
+    train_cfg=dict(
+        mix_loss=dict(
```
Comment (JihwanEom) [inline models/segmentation/_base_/ocr_litehrnet18_mod2.py]: Could you check the operation of mix_loss? I presume we might not need this option because we are using only one loss.

File: models/segmentation/_base_/ocr_litehrnet_s_mod2.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+             enable_aggregator=True,
+             aggregator_merge_norm=None,
+             aggregator_use_concat=False,
+             enable_out_norm=False,
+             enable_loss_equalizer=True,
+             loss_decode=[
+                 dict(type='CrossEntropyLoss',
+                      use_sigmoid=False,
+                      sampler=dict(type='MaxPoolingPixelSampler', ratio=0.25, p=1.7),
+                      loss_weight=4.0),
```
Comment (JihwanEom) [inline models/segmentation/_base_/ocr_litehrnet_s_mod2.py]: This loss_weight is also suspicious.

File: models/segmentation/_base_/ocr_litehrnet_x_mod3.py [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+            weight=0.1
+        ),
+        loss_reweighting=dict(
+            weights={'decode_0.loss_seg': 0.9},
+            momentum=0.1
+        ),
+    ),
+    test_cfg=dict(
+        mode='whole',
+        output_scale=10.0,
```
Comment (JihwanEom) [inline models/segmentation/_base_/ocr_litehrnet_x_mod3.py]: All models (18-mod2, s-mod2, x-mod3) are using same output_scale in config, but I'm not sure this is right config for each architecture. 
Could you check whether output_scale config is working correctly?

File: mpa/modules/models/classifiers/sam_classifier.py [hunk 9]
```diff
@@ -82,6 +83,7 @@ def forward_train(self, img, gt_label, **kwargs):
         if self.multilabel:
             loss = self.head.forward_train(x, gt_label, **kwargs)
         else:
+            gt_label = torch.squeeze(gt_label)
```
Comment (JihwanEom) [inline mpa/modules/models/classifiers/sam_classifier.py]: How about `gt_label.squeeze()` to prevent importing additional dependency.

File: recipes/stages/_base_/logs/tensorboard_logger.py [hunk 10]
```diff
@@ -4,7 +4,7 @@
 log_config = dict(
     interval=100,
     hooks=[
-        dict(type='TextLoggerHook'),
+        dict(type='TextLoggerHook', ignore_last=False),
```
Comment (JihwanEom) [inline recipes/stages/_base_/logs/tensorboard_logger.py:7]: I'm thinking about more intuitive name for this.. how about `force_print=True`

File: recipes/stages/segmentation/train.py [hunk 11]
```diff
@@ -12,7 +12,16 @@
     weight_decay=0.0001,
 )
 
-lr_config = dict(min_lr=1e-06)
+lr_config = dict(
+    policy='ReduceLROnPlateau',
+    metric='mIoU',
```
Comment (JihwanEom) [inline recipes/stages/segmentation/train.py]: Is the current segmentation task is looking 'mIoU' to return the best checkpoint? I'm confused with 'mDice' metric.

File: recipes/stages/_base_/logs/tensorboard_logger.py (hunk 10 above)
Comment (supersoob) [reply recipes/stages/_base_/logs/tensorboard_logger.py:7]: This is a built-in parameter from mmcv. I found that it has supported for logging last_iter.

File: recipes/stages/segmentation/train.py (hunk 11 above)
Comment (chuneuny-emily) [reply recipes/stages/segmentation/train.py]: It should be changed to mDice. I changed it.

File: recipes/stages/segmentation/train.py [hunk 12]
```diff
@@ -23,15 +32,15 @@
...
                  r'\w*[.]?decode_head\.', r'\w*[.]?auxiliary_head\.']
 )
 
 custom_hooks = [
     dict(type='LazyEarlyStoppingHook',
-         patience=5,
-         iteration_patience=1000,
+         patience=8,
+         iteration_patience=0,
          metric='mIoU',
```
Comment (JihwanEom) [inline recipes/stages/segmentation/train.py]: @chuneuny-emily then, I think this metric also should be changed with 'mDice', am I right?

Comment (chuneuny-emily) [reply recipes/stages/segmentation/train.py]: Yes. I changed it too.

File: recipes/stages/segmentation/class_incr.py [hunk 13]
```diff
@@ -5,10 +5,11 @@
 ]
 
 optimizer = dict(
-    type='SGD',
+    _delete_=True,
+    type='Adam',
```
Comment (JihwanEom) [inline recipes/stages/segmentation/class_incr.py:9]: Adam shows better results than SGD? I remember you said SGD is better than Adam.

File: models/segmentation/_base_/ocr_litehrnet_x_mod3.py (hunk 8 above)
Comment (chuneuny-emily) [reply models/segmentation/_base_/ocr_litehrnet_x_mod3.py]: That configs are the default config on mmseg. Is there any reason you think that is not right config for each architecture?

File: recipes/stages/segmentation/class_incr.py (hunk 13 above)
Comment (chuneuny-emily) [reply recipes/stages/segmentation/class_incr.py:9]: SGD gave slight accuracy gain w/ 120 image set, but it affected too slow training time and accuracy drop on large backbone w/ 12 image set. With the small set, Adam is giving very fast learning w/ slight accuracy drop.

File: models/segmentation/_base_/ocr_litehrnet_x_mod3.py (hunk 8 above)
Comment (JihwanEom) [reply models/segmentation/_base_/ocr_litehrnet_x_mod3.py]: When I tested the performance of segmentation task, I remember output_scale had huge effect.
I'm not sure this `output_scale=10.0` is right config without OCR head and also wondering about whether this option is working correctly.

File: models/segmentation/_base_/ocr_litehrnet_s_mod2.py (hunk 7 above)
Comment (chuneuny-emily) [reply models/segmentation/_base_/ocr_litehrnet_s_mod2.py]: 4.0 is default value on mmseg. Is there any reason you think it's suspicious?

File: recipes/stages/segmentation/class_incr.py (hunk 13 above)
Comment (JihwanEom) [reply recipes/stages/segmentation/class_incr.py:9]: Okay, thank you for detailed explanation. Could you confirm grad_norm for segmentation tasks?
I remember OTEseg team applied Adam optimizer to stabilize grad_norm of very early epoch.

File: models/segmentation/_base_/ocr_litehrnet_s_mod2.py (hunk 7 above)
Comment (JihwanEom) [reply models/segmentation/_base_/ocr_litehrnet_s_mod2.py]: Yes, I think this can cause unstable training by multiplying 4 to origin loss value. In OTEseg s_mod2 model, it is using 'CrossEntropyLoss', 'GeneralizedDiceLoss', 'AMSoftmaxLoss', each weight is 4.0, 4.0, 1.0 respectively. 
So, I'm suspicious `loss_weight=4.0` can be too large without OCR head/Aux head.

File: recipes/stages/segmentation/class_incr.py (hunk 13 above)
Comment (chuneuny-emily) [reply recipes/stages/segmentation/class_incr.py:9]: grad_norm works w/ Adam, and does not work w/ SGD.

File: models/segmentation/_base_/ocr_litehrnet_s_mod2.py (hunk 7 above)
Comment (chuneuny-emily) [reply models/segmentation/_base_/ocr_litehrnet_s_mod2.py]: Thanks for your feedback. Since 'GeneralizedDiceLoss', 'AMSoftmaxLoss' were removed, I changed loss_weight to 1. I did sanity check and no effect on accuracy.

File: models/segmentation/_base_/ocr_litehrnet18_mod2.py (hunk 6 above)
Comment (chuneuny-emily) [reply models/segmentation/_base_/ocr_litehrnet18_mod2.py]: Thanks for your feedback. Since there is only CE loss, mix_loss is not required. I removed it.

File: models/segmentation/_base_/ocr_litehrnet18_mod2.py (hunk 5 above)
Comment (chuneuny-emily) [reply models/segmentation/_base_/ocr_litehrnet18_mod2.py]: Thanks for your feedback. Since there is only CE loss, mix_loss is not required. I removed it.

File: models/segmentation/_base_/ocr_litehrnet_x_mod3.py (hunk 8 above)
Comment (chuneuny-emily) [reply models/segmentation/_base_/ocr_litehrnet_x_mod3.py]: It is working, if I remove it, on large, new class was not learned. About the value 10, I'll conduct the experiments whether it's proper default per model.

File: models/segmentation/_base_/ocr_litehrnet18_mod2.py (hunk 4 above)
Comment (chuneuny-emily) [reply models/segmentation/_base_/ocr_litehrnet18_mod2.py]: I updated it.

File: mpa/modules/models/segmentors/class_incr_segmentor.py [hunk 14]
```diff
@@ -45,9 +45,8 @@ def load_state_dict_pre_hook(model, model_classes, chkpt_classes, chkpt_dict, pr
 
         model_dict = model.state_dict()
         param_names = [
-            'decode_head.0.conv_seg.weight',
-            'decode_head.0.conv_seg.bias',
-            'decode_head.1.conv_seg.weight',
+            'decode_head.conv_seg.weight',
+            'decode_head.conv_seg.bias',
```
Comment (goodsong81) [inline mpa/modules/models/segmentors/class_incr_segmentor.py:49]: Is this enough for arch change?
Can we load old OTE model weights from, for example, SC1.1 model with OCR head?

**Review by goodsong81 — APPROVED** [review]

File: mpa/modules/models/segmentors/class_incr_segmentor.py (hunk 14 above)
Comment (chuneuny-emily) [reply mpa/modules/models/segmentors/class_incr_segmentor.py:49]: Currently it does not support. Let me update it on next PR.
